### PR TITLE
Refactor more handlers to Kotlin

### PR DESCRIPTION
**Pull request by dpgraham**

### CODE REVIEW

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/Keys.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.ActionType.KEY_UP
+import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.InputSourceType.KEY
+
+class Keys : RequestHandler<TextParams, Void?> {
+
+    @Throws(AppiumException::class)
+    override fun handle(params: TextParams): Void? {
+        val runnable = UiControllerRunnable<Void> { uiController ->
+            // Add a list of keyDown + keyUp actions for each key
+            val keyActions = arrayListOf<Action>()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/Keys.kt]: use map

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        throw InvalidArgumentException("Target must not be empty and must be of type: 'activity', 'application'")
+
+    }
+
+    @Throws(AppiumException::class)
+    private fun invokeBackdoorMethods(invokeOn: Any, ops: List<InvocationOperation>): Any? {
+        var invocationResult: Any? = null
+        var invocationTarget: Any? = invokeOn
+        ops.forEach {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt]: can reduce be used here?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileViewFlash.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        } catch (e: InterruptedException) {
+            throw AppiumException(e)
+        }
+
+        return null
+    }
+
+    companion object {
+
+        private val DURATION_MILLIS = 30
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileViewFlash.kt]: do we need this companion object?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/PointerEventHandler.kt [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+    @Throws(AppiumException::class)
+    private fun handleTouchScroll(params: MotionEventParams) {
+        // Fabricate a scroll event
+        // Do halfway points, by default
+        var startX = displayMetrics.widthPixels / 2 - params.x / 2
+        var startY = displayMetrics.heightPixels / 2 - params.y / 2
+
+        params.elementId?.let {
+            val view = Element.getViewById(params.elementId)
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/PointerEventHandler.kt]: it

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ReleaseActions.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import io.appium.espressoserver.lib.viewaction.UiControllerPerformer
+import io.appium.espressoserver.lib.viewaction.UiControllerRunnable
+
+
+class ReleaseActions : RequestHandler<AppiumParams, Void?> {
+
+    @Throws(AppiumException::class)
+    override fun handle(params: AppiumParams): Void? {
+
+        val runnable = UiControllerRunnable<Void> {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ReleaseActions.kt]: might be way cleaner with lambdas

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScrollToPage.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        val viewInteraction = Element.getViewInteractionById(params.elementId)
+        try {
+            val smoothScroll = params.smoothScroll
+            params.scrollTo?.let {
+                val scrollAction = when (it) {
+                    ScrollToPageParams.ScrollTo.FIRST -> ViewPagerActions.scrollToFirst(smoothScroll)
+                    ScrollToPageParams.ScrollTo.LAST -> ViewPagerActions.scrollToLast(smoothScroll)
+                    ScrollToPageParams.ScrollTo.LEFT -> ViewPagerActions.scrollLeft(smoothScroll)
+                    ScrollToPageParams.ScrollTo.RIGHT -> ViewPagerActions.scrollRight(smoothScroll)
+                }
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScrollToPage.kt]: de we need else?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        val view = Element.getViewById(id)
+
+        // Convert the array of text to a String
+        val textArray = params.value
+        val stringBuilder = StringBuilder()
+        for (text in textArray) {
+            stringBuilder.append(text)
+        }
+
+        var value: String? = stringBuilder.toString()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt]: reduce?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        } catch (e: PerformException) {
+            throw InvalidElementStateException("sendKeys", params.elementId, e)
+        } catch (e: RuntimeException) {
+            e.message?.let {
+                it.contains("IME does not understand how to translate") ?: throw e
+            }
+            params.text?.let {
+                value = it
+            }
+            logger.debug(String.format("Trying replaceText action as a workaround " + "to type the '%s' text into the input field", value))
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt]: Use Kotlin string template

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileViewFlash.kt [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        val duration = params.durationMillis ?: DURATION_MILLIS
+        val repeatCount = params.repeatCount ?: REPEAT_COUNT
+
+        val view = Element.getViewById(params.elementId)
+        val latch = CountDownLatch(1)
+        InstrumentationRegistry.getInstrumentation().runOnMainSync {
+            val animation = AlphaAnimation(1f, 0f)
+            animation.repeatMode = Animation.REVERSE
+            animation.duration = duration.toLong()
+            animation.setRepeatCount(repeatCount)
```
Comment (KazuCocoa) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileViewFlash.kt]: Can use `animation.repeatCount = repeatCount` style

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/Keys.kt (hunk 1 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/Keys.kt]: I wanted to do that, but the problem is that there's two values being added on each iteration.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScrollToPage.kt (hunk 6 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScrollToPage.kt]: We don't, actually.

Kotlin `when` statements only compile if
1. They're "exhaustive" (ie: every pathway is handled). 
1. They have an `else` statement to cover the default path

In this case, it's exhaustive so the else statement will be omitted. If another `ScrollTo` param is added, compilation will fail and we'll have to add it.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt (hunk 7 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt]: Found an even better way: `var value: String? = textArray.joinToString(seperator="")`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt (hunk 8 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/SendKeys.kt]: I wish the code converter would cover that.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt (hunk 2 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt]: I used `fold` which is similar to `reduce`.

Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.kt]: Works like `collection.fold(<initValue>){ value, item -> (do something) }`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ReleaseActions.kt (hunk 5 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ReleaseActions.kt]: I'm going to leave a `TODO` for this.... I'll need to understand the lambda -> runnable relation better first. In the future, I can do a PR to convert _all_ runnables to lambda.

**Pull request merged**